harness/tests/cli/test_auth.py

```diff
@@ -29,74 +34,502 @@
-            assert authentication.session.username == user or "determined"
-            assert authentication.session.token == (
-                "det.token" if user == "determined" else "bob.token"
+    def getpass(*_: Any) -> str:
+        return "prompt_pass"
+
+    def _is_token_valid(master_url: str, token: str, cert: Any) -> bool:
+        return token in ["cache", "env_token"]
+
+    def do_login(*_: Any) -> str:
+        return "new"
+
+    mock_getpass.side_effect = getpass
+    mock_is_token_valid.side_effect = _is_token_valid
+    mock_do_login.side_effect = do_login
+
+    for scenario in scenario_set.scenarios():
+        with contextlib.ExitStack() as es:
+            # Configure environment.
+            es.enter_context(setenv_optional("DET_USER", scenario.env_user))
+            es.enter_context(setenv_optional("DET_PASS", scenario.env_pass))
+            es.enter_context(setenv_optional("DET_USER_TOKEN", scenario.env_token))
+
+            # Configure allowable TokenStore calls.
+            mts = es.enter_context(MockTokenStore(strict=False))
+            mts.get_token("user", retval=scenario.cache)
+            mts.get_token("extra", retval=scenario.cache)
+            mts.get_token("determined", retval=scenario.cache)
+            mts.drop_user("user")
+            mts.drop_user("extra")
+            mts.set_token("user", "new")
+            mts.set_token("determined", "new")
+
+            # The cache active user is always "user".  This deserves some explanation.  Effectively,
+            # since this is only called within the `default_load_user_password()` function, that
+            # function will never show evidence that it checked the environment.  However, for the
+            # remainder of the authentication flow, it never matters again that the user came from
+            # the TokenStore.get_active_user(), so if we modeled that in the table it wouldn't
+            # actually increase code coverage for the Authentication object, which is what we are
+            # focusing on.
+            mts.get_active_user(retval=scenario.cache and "user")
+
+            try:
+                auth = authentication.Authentication(
+                    "master_url", scenario.req_user, scenario.req_pass, None
+                )
+
+                # Make sure we got the results we expected.
+                for result in scenario_set.results:
+                    if isinstance(result, Check):
+                        mock_is_token_valid.assert_has_calls(
+                            [mock.call("master_url", result.token, None)]
+                        )
+                    elif isinstance(result, DoLogin):
+                        mock_do_login.assert_has_calls(
+                            [mock.call("master_url", result.username, result.password, None)]
+                        )
+                    elif isinstance(result, Use):
+                        assert auth.session.token == result.token
```

Comment:
We can add `assert auth.session.username == result.user` here if we implement some of the test scenario separation changes that I suggested above!